drivers/i3c/i3c_stm32.c

```diff
@@ -0,0 +1,2163 @@
+
+	if (i2c_bus_freq != 0) {
+		if (i2c_bus_freq > 400000) {
+			/* Mixed bus with I2C FM+ device */
+			free_timing = (uint8_t)ceil(
+				(STM32_I3C_TBUF_FMP_MIN_NS * i3c_clock / 1e9 - 0.5) / 2);
+		} else {
+			/* Mixed bus with I2C FM device */
+			free_timing = (uint8_t)ceil(
+				(STM32_I3C_TBUF_FM_MIN_NS * i3c_clock / 1e9 - 0.5) / 2);
+		}
+	} else {
+		if (config->drv_cfg.dev_list.num_i2c > 0) {
+			enum i3c_bus_mode mode = i3c_bus_mode(&config->drv_cfg.dev_list);
+
+			if (mode == I3C_BUS_MODE_MIXED_FAST) {
+				if (get_i3c_lvr_ic_mode(&config->drv_cfg.dev_list) ==
+				    I3C_LVR_I2C_FM_MODE) {
+					/* Mixed bus with I2C FM device */
+					free_timing = (uint8_t)ceil(
+						(STM32_I3C_TBUF_FM_MIN_NS * i3c_clock / 1e9 - 0.5) /
+						2);
+				} else {
+					/* Mixed bus with I2C FM+ device */
+					free_timing = (uint8_t)ceil(
+						(STM32_I3C_TBUF_FMP_MIN_NS * i3c_clock / 1e9 -
+						 0.5) /
+						2);
+				}
+			} else {
+				return -EINVAL;
+			}
+		} else {
+			/* Pure I3C bus */
+			free_timing =
+				(uint8_t)ceil((STM32_I3C_TCAS_MIN_NS * i3c_clock / 1e9 - 0.5) / 2);
+		}
+	}
+
+	aval = DIV_ROUND_UP(1000ull * i3c_clock, 1000000000ull) - 1;
+
+	LL_I3C_SetDataHoldTime(i3c, LL_I3C_SDA_HOLD_TIME_1_5);
+	LL_I3C_SetFreeTiming(i3c, free_timing);
+	LL_I3C_SetControllerActivityState(i3c, LL_I3C_OWN_ACTIVITY_STATE_0);
+	LL_I3C_SetAvalTiming(i3c, aval);
+
+	LOG_DBG("TimingReg1 = 0x%08x", LL_I3C_GetCtrlBusCharacteristic(i3c));
+
+	return 0;
+}
+
+/* Configures the I3C module in controller mode */
+static int i3c_stm32_configure(const struct device *dev, enum i3c_config_type type, void *cfg)
+{
+	int ret;
+
+	if (type == I3C_CONFIG_TARGET || type == I3C_CONFIG_CUSTOM) {
+		return -ENOTSUP;
+	}
+
+	struct i3c_stm32_data *data = dev->data;
+	struct i3c_config_controller *ctrl_cfg = cfg;
+
+	if ((ctrl_cfg != NULL) && ((ctrl_cfg->scl.i2c == 0U) || (ctrl_cfg->scl.i3c == 0U))) {
+		return -EINVAL;
+	}
+
+	if (ctrl_cfg != NULL) {
+		data->drv_data.ctrl_config.scl.i3c = ctrl_cfg->scl.i3c;
+		data->drv_data.ctrl_config.scl.i2c = ctrl_cfg->scl.i2c;
+	}
+
+	ret = i3c_stm32_activate(dev);
+	if (ret != 0) {
+		LOG_ERR("Clock and GPIO could not be initialized for the I3C module, err=%d", ret);
+		return ret;
+	}
+
+	ret = i3c_stm32_config_clk_wave(dev);
+	if (ret != 0) {
+		LOG_ERR("TimigReg0 timing could not be calculated, err=%d", ret);
+		return ret;
+	}
+
+	ret = i3c_stm32_config_ctrl_bus_char(dev);
+	if (ret != 0) {
+		LOG_ERR("TimingReg1 timing could not be calculated, err=%d", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int i3c_stm32_i2c_configure(const struct device *dev, uint32_t config)
+{
+	struct i3c_stm32_data *data = dev->data;
+	struct i3c_config_controller *ctrl_config = &data->drv_data.ctrl_config;
+
+	switch (I2C_SPEED_GET(config)) {
+	case I2C_SPEED_FAST:
+		ctrl_config->scl.i2c = 400000;
+		break;
+	case I2C_SPEED_FAST_PLUS:
+		ctrl_config->scl.i2c = 1000000;
+		break;
+	default:
+		return -EINVAL;
+	}
```

Comment:
Are other speeds not supported such as `I2C_SPEED_STANDARD`?

Reply:
The only 2 I2C speeds supported for this chip are the FM and the FM+